src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3070,380 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+		CopyConnectionState *connectionState = NULL;
+		CopyPlacementState *placementState = NULL;
+		
+		MultiConnection *connection =
+			CopyGetPlacementConnection(placement, connectionStateHash, stopOnFailure);
+		if (connection == NULL)
+		{
+			failedPlacementCount++;
+			continue;
+		}
+
+		connectionState = GetConnectionState(connectionStateHash, connection);
+		placementState = palloc0(sizeof(CopyPlacementState));
+		placementState->shardState = shardState;
+		placementState->data = makeStringInfo();
+		placementState->connectionState = connectionState;
+		dlist_push_head(&connectionState->bufferedPlacementList,
+						&placementState->bufferedPlacementNode);
+		shardState->placementStateList = lappend(shardState->placementStateList,
+												 placementState);
+	}
+
+	/* if all placements failed, error out */
+	if (failedPlacementCount == list_length(finalizedPlacementList))
+	{
+		ereport(ERROR, (errmsg("could not connect to any active placements")));
+	}
+
+	/*
+	 * If stopOnFailure is true, we just error out and code execution should
+	 * never reach to this point. This is the case for reference tables and
+	 * copy from worker nodes.
+	 */
+	Assert(!stopOnFailure || failedPlacementCount == 0);
+
+	MemoryContextReset(localContext);
+}
+
+
+/*
+ * CopyGetPlacementConnection assigns a connection to the given placement. If
+ * a connection has already been assigned the placement in the current transaction
+ * then it reuses the connection. Otherwise, it requests a connection for placement.
+ */
+static MultiConnection *
+CopyGetPlacementConnection(ShardPlacement *placement, HTAB *connectionStateHash,
+						   bool stopOnFailure)
+{
+	MultiConnection *connection = NULL;
+	uint32 connectionFlags = FOR_DML;
+	char *nodeUser = CurrentUserName();
+	ShardPlacementAccess *placementAccess = NULL;
+
+	/*
+	 * Determine whether the task has to be assigned to a particular connection
+	 * due to a preceding access to the placement in the same transaction.
+	 */
+	placementAccess = CreatePlacementAccess(placement, PLACEMENT_ACCESS_DML);
+	connection = GetPlacementListConnectionIfCached(connectionFlags,
+													list_make1(placementAccess),
+													NULL);
+	if (connection != NULL)
+	{
+		return connection;
+	}
+
+	/* if execution mode set to sequential, reuse worker connection if available */
+	if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+	{
+		connection = GetWorkerConnectionIfCached(placement->nodeName,
+												 placement->nodePort,
+												 connectionStateHash);
+		if (connection != NULL)
+		{
+			return connection;
+		}
+	}
+
+	/*
+	 * For placements that haven't been assigned a connection by a previous command
+	 * in the current transaction, we use a separate connection per placement for
+	 * hash-distributed tables in order to get the maximum performance.
+	 */
+	if (placement->partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		connectionFlags |= CONNECTION_PER_PLACEMENT;
```

Comment:
I guess we should not use connection per placement for sequential mode. We're setting the flag, but, we're still using the same connection to COPY multiple shards. 

Probably not a big deal, just while reading that I felt a bit unusual.

Reply:
For context: The CONNECTION_PER_PLACEMENT flag solved a specific problem that, after this PR, no longer really exists.

e.g. if distribution column values 1 and 2 are in different shards on the same node and we did not use CONNECTION_PER_PLACEMENT:
```sql
BEGIN;
INSERT INTO test VALUES (2,3);
\COPY test FROM STDIN WITH CSV
1,2 -- uses the already opened connection that was used for the insert
2,3 -- needs to use the connection that was used for the insert, but that's now in use
\.
```

The same problem existed for multi-shard queries and DML in the old executors. 

Going forward, we could consider removing the flag but it might still be beneficial to use it for COPY because it would avoid multi-plexing in some scenarios.

w.r.t. the sequential case I think neither using connection exclusively, nor the CONNECTION_PER_PLACEMENT flag could work.